File: 11/11286.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <cmath>
using namespace std;

struct compare {
	bool operator()(int a, int b) {
		if (abs(a) == abs(b)) return a > b;
		return abs(a) > abs(b);
	}
};

int main() {
	int N; scanf("%d", &N);	// 연산의 개수
	priority_queue<int, vector<int>, compare> heap;	// 최대 힙

	for (int i = 0; i < N; i++) {
		int x; scanf("%d", &x);

		if (x == 0) {
			if (heap.empty()) printf("0\n");
			else {
				printf("%d\n", heap.top());
				heap.pop();
			}
		}
		else heap.push(x);
	}
}

File: 11/11286.java
import java.util.*;

public class Main {

    public static Scanner scan = new Scanner(System.in);

    public static void main(String[] args){
        Comparator<Integer> comparator = new Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                if(Math.abs(a) == Math.abs(b)){
                    if(a == b) return 0;
                    else if(a < b) return -1;
                    else return 1;
                }
                else if(Math.abs(a) < Math.abs(b)) return -1;
                else return 1;
            }
        };

        int N = scan.nextInt();
        PriorityQueue<Integer> heap = new PriorityQueue<>(comparator);

        for(int i=0; i<N; i++){
            int x = scan.nextInt();

            if(x == 0){
                if(heap.isEmpty()) System.out.println(0);
                else System.out.println(heap.poll());
            }
            else heap.offer(x);
        }
    }
}
